Incompatible with Gradle 4?

Since I'm using Gradle Kotlin Script I use the following Gradle distribution:
```
https://repo.gradle.org/gradle/dist-snapshots/gradle-script-kotlin-4.0-20170506114645+0000-all.zip
```

When using that version of Gradle, the versions plugin does no longer work:

```
$ ./gradlew dependencyUpdates --stacktrace
:dependencyUpdates FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dependencyUpdates'.
> java.lang.String cannot be cast to org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.Version

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':dependencyUpdates'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:60)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:247)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:297)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:289)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:166)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:238)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:227)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:121)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:77)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:102)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:96)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:586)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:541)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:96)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.Version
        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.StaticVersionComparator.compare(StaticVersionComparator.java:25)
        at java_util_Comparator$compare.call(Unknown Source)
        at com.github.benmanes.gradle.versions.updates.VersionComparator.compare(VersionComparator.groovy:49)
        at com.github.benmanes.gradle.versions.updates.VersionComparator$compare.call(Unknown Source)
        at com.github.benmanes.gradle.versions.updates.Resolver.<init>(Resolver.groovy:53)
        at com.github.benmanes.gradle.versions.updates.DependencyUpdates$_run_closure2.doCall(DependencyUpdates.groovy:57)
        at com.github.benmanes.gradle.versions.updates.DependencyUpdates.run(DependencyUpdates.groovy:56)
        at com.github.benmanes.gradle.versions.updates.DependencyUpdates$run.call(Unknown Source)
        at com.github.benmanes.gradle.versions.updates.DependencyUpdatesTask.dependencyUpdates(DependencyUpdatesTask.groovy:54)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:139)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:132)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:706)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:689)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:297)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:289)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:166)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 27 more


BUILD FAILED in 1s
1 actionable task: 1 executed, 0 avoided (0%)
```

Having the same issue when using Gradle Script Kotlin 0.9, which based on 4.0. So is there any workaround I can apply in the mean time? 

```
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.Version
        at org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.StaticVersionComparator.compare(StaticVersionComparator.java:25)
```

--- a/README.md
+++ b/README.md
@@ -12,7 +12,7 @@ add it to your top-level build script using the following configuration:
 
 ```groovy
 plugins {
-    id 'com.github.ben-manes.versions' version '0.14.0'
+  id 'com.github.ben-manes.versions' version '0.15.0'
 }
 ```
 or via the
@@ -27,12 +27,12 @@ buildscript {
   }
 
   dependencies {
-    classpath 'com.github.ben-manes:gradle-versions-plugin:0.14.0'
+    classpath 'com.github.ben-manes:gradle-versions-plugin:0.15.0'
   }
 }
 ```
 
-The current version is known to work with Gradle versions up to 3.4.
+The current version is known to work with Gradle versions up to 4.0-rc1.
 
 ## Tasks
 

--- a/build.gradle
+++ b/build.gradle
@@ -17,7 +17,7 @@ apply plugin: 'nexus'
 apply plugin: 'codenarc'
 
 group = 'com.github.ben-manes'
-version = '0.14.0'
+version = '0.15.0'
 
 sourceCompatibility = '1.6'
 

--- a/src/main/groovy/com/github/benmanes/gradle/versions/updates/VersionComparator.groovy
+++ b/src/main/groovy/com/github/benmanes/gradle/versions/updates/VersionComparator.groovy
@@ -29,12 +29,10 @@ import static groovy.transform.TypeCheckingMode.SKIP
  */
 @TypeChecked
 class VersionComparator implements Comparator<String> {
-  static final String GRADLE_24 =
-    'org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.DefaultVersionComparator'
-  static final String GRADLE_23 =
-    'org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.StaticVersionComparator'
-  static final String GRADLE_18 =
-    'org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy.ResolverStrategy'
+  static final String BASE_PKG = 'org.gradle.api.internal.artifacts.ivyservice.ivyresolve.strategy'
+  static final String GRADLE_24 = BASE_PKG + '.DefaultVersionComparator'
+  static final String GRADLE_23 = BASE_PKG + '.StaticVersionComparator'
+  static final String GRADLE_18 = BASE_PKG + '.ResolverStrategy'
   static final String GRADLE_10 =
     'org.gradle.api.internal.artifacts.version.LatestVersionSemanticComparator'
 
@@ -64,7 +62,7 @@ class VersionComparator implements Comparator<String> {
   @TypeChecked(SKIP)
   private List<Closure<Comparator<String>>> candidates() {
     return [
-      { createInstance(GRADLE_24).asStringComparator() },
+      { createInstance(GRADLE_24); makeStringComparator() },
       { createInstance(GRADLE_23) },
       { createInstance(GRADLE_18).getVersionMatcher() },
       { createInstance(GRADLE_10) }
@@ -75,5 +73,18 @@ class VersionComparator implements Comparator<String> {
   Closure<Object> createInstance = { String className ->
     def classLoader = Thread.currentThread().getContextClassLoader()
     return classLoader.loadClass(className).newInstance()
+  }
+
+  @TypeChecked(SKIP)
+  private Comparator<String> makeStringComparator() {
+    def baseComparator = createInstance(BASE_PKG + '.StaticVersionComparator')
+    def versionParser = createInstance(BASE_PKG + '.VersionParser')
+    return new Comparator<String>() {
+      public int compare(String string1, String string2) {
+        return baseComparator.compare(
+          versionParser.transform(string1),
+          versionParser.transform(string2));
+      }
+    }
   }
 }

--- a/src/test/groovy/com/github/benmanes/gradle/versions/DifferentGradleVersionsSpec.groovy
+++ b/src/test/groovy/com/github/benmanes/gradle/versions/DifferentGradleVersionsSpec.groovy
@@ -79,7 +79,8 @@ public class DifferentGradleVersionsSpec extends Specification {
       '2.0',
       '3.3',
       '3.4',
-      '3.5'
+      '3.5',
+      '4.0-rc-1'
     ]
   }
 }

